data/transactions/logic/assembler_test.go

```diff
@@ -2621,6 +2621,29 @@ func TestScratchTypeCheck(t *testing.T) {
 	testProg(t, "callsub A; load 0; btoi; return; A: byte 0x01; store 0; retsub", AssemblerMaxVersion)
 	// But the scratchspace should still be tracked after the callsub
 	testProg(t, "callsub A; int 1; store 0; load 0; btoi; return; A: retsub", AssemblerMaxVersion, Expect{1, "btoi arg 0..."})
+
+}
+
+func TestScratchBounds(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	os := testProg(t, "int 5; store 1; load 1; return;", AssemblerMaxVersion)
+	sv := os.known.scratchSpace[1]
+	require.Equal(t, sv.AVMType, avmUint64)
+	require.ElementsMatch(t, sv.Bound, static(5))
+
+	os = testProg(t, "int 5; store 1; load 1; int 1; int 1; stores; return;", AssemblerMaxVersion)
+	sv = os.known.scratchSpace[1]
```

Comment:
Is typeStores so smart that only scratchSpace[1] is affected here? (since you have a bounded argument to `stores`?)

If so, it could be so smart that it _replaces_ scratch[1] with the type it has, instead of union, because you know _that_ slot is being replaced.

If not, let's show that in the test, by showing that the bound on scratch[4] is also changed.

Reply:
It is now smart enough IF the stack value of the index is a const, I've updated the test to reflect this